File: cpp/src/ev2/ut6_pooavanzadaestructuras/u01herenciapolimorfismo/practicas/madridcesjuanpablosegundotrabajadores/trabajadores_resuelto.cpp
#include <iostream>
#include <vector>
#include <memory>
#include <string>
using namespace std;
class Trabajador { public: string nombre,apellido,dni; double sueldo;
  Trabajador(string n,string a,string d,double s):nombre(n),apellido(a),dni(d),sueldo(s){}
  virtual ~Trabajador()=default; virtual double calcularSalarioMensual()=0;
  virtual void mostrarDatos(){ cout<<"nombre = "<<nombre<<"\ndni = "<<dni<<"\nsalario mensual = "<<calcularSalarioMensual()<<"\n"; } };
class Asalariado: public Trabajador { public: int pagas; double cuota;
  Asalariado(string n,string a,string d,double s,int p,double c):Trabajador(n,a,d,s),pagas(p),cuota(c){}
  double calcularSalarioMensual() override { return (sueldo*0.79)/pagas; }
  void mostrarDatos() override { cout<<"Tipo = Asalariado\n"; Trabajador::mostrarDatos(); } };
class Autonomo: public Trabajador { public: double cuota;
  Autonomo(string n,string a,string d,double s,double c):Trabajador(n,a,d,s),cuota(c){}
  double calcularSalarioMensual() override { return (sueldo-12*cuota)/12; }
  void mostrarDatos() override { cout<<"Tipo = Autonomo\n"; Trabajador::mostrarDatos(); } };
class Jefe: public Trabajador { public: int acciones; double beneficio;
  Jefe(string n,string a,string d,double s,int ac,double b):Trabajador(n,a,d,s),acciones(ac),beneficio(b){}
  double calcularSalarioMensual() override { return sueldo/12; }
  void mostrarDatos() override { cout<<"Tipo = Jefe\n"; Trabajador::mostrarDatos(); } };
int main(){ vector<unique_ptr<Trabajador>> lista;
  lista.push_back(make_unique<Asalariado>("Ana","G","1A",24000,14,100));
  lista.push_back(make_unique<Autonomo>("Luis","P","2B",30000,200));
  lista.push_back(make_unique<Jefe>("Eva","R","3C",60000,10,1000));
  for(auto& t: lista) t->mostrarDatos(); return 0; }

File: js/ev2/ut5-poo-colecciones/03-objetos-y-clases/practicas/madrid-cesjuanpablosegundo-trabajadores/trabajadores-resuelto.js
class Trabajador { constructor(nombre,apellido,dni,sueldo){ this.nombre=nombre; this.apellido=apellido; this.dni=dni; this.sueldo=sueldo; }
  calcularSalarioMensual(){ throw new Error("abstract"); }
  mostrarDatos(){ console.log(`nombre = ${this.nombre}`); console.log(`dni = ${this.dni}`); console.log(`salario mensual = ${this.calcularSalarioMensual()}`); } }
class Asalariado extends Trabajador { constructor(n,a,d,s,pagas,cuota){ super(n,a,d,s); this.pagas=pagas; this.cuota=cuota; }
  calcularSalarioMensual(){ return (this.sueldo*0.79)/this.pagas; } mostrarDatos(){ console.log("Tipo = Asalariado"); super.mostrarDatos(); } }
class Autonomo extends Trabajador { constructor(n,a,d,s,cuota){ super(n,a,d,s); this.cuota=cuota; }
  calcularSalarioMensual(){ return (this.sueldo-12*this.cuota)/12; } mostrarDatos(){ console.log("Tipo = Autonomo"); super.mostrarDatos(); } }
class Jefe extends Trabajador { constructor(n,a,d,s,acciones,beneficio){ super(n,a,d,s); this.acciones=acciones; this.beneficio=beneficio; }
  calcularSalarioMensual(){ return this.sueldo/12; } mostrarDatos(){ console.log("Tipo = Jefe"); super.mostrarDatos(); } }
const lista = [new Asalariado("Ana","G","1A",24000,14,100), new Autonomo("Luis","P","2B",30000,200), new Jefe("Eva","R","3C",60000,10,1000)];
lista.forEach(t => t.mostrarDatos());
